libp2p/protocols/pubsub/gossipsub/behavior.nim

```diff
@@ -310,22 +310,31 @@ proc rebalanceMesh*(g: GossipSub, topic: string, metrics: ptr MeshMetrics = nil)
   var
     prunes, grafts: seq[PubSubPeer]
     npeers = g.mesh.peers(topic)
+    defaultMesh: HashSet[PubSubPeer]
+    backingOff = g.backingOff.getOrDefault(topic)
 
   if npeers  < g.parameters.dLow:
     trace "replenishing mesh", peers = npeers
     # replenish the mesh if we're below Dlo
-    var candidates = toSeq(
-      g.gossipsub.getOrDefault(topic, initHashSet[PubSubPeer]()) -
-      g.mesh.getOrDefault(topic, initHashSet[PubSubPeer]())
-    ).filterIt(
-      it.connected and
-      # avoid negative score peers
-      it.score >= 0.0 and
-      # don't pick explicit peers
-      it.peerId notin g.parameters.directPeers and
-      # and avoid peers we are backing off
-      it.peerId notin g.backingOff.getOrDefault(topic)
-    )
+
+    # This is a workaround to get the HashSet from the table
+    # without creating a copy.
+    proc getCandidates(currentMesh: var HashSet[PubSubPeer]): seq[PubSubPeer] =
+      g.gossipSub.withValue(topic, peerList):
+        for it in peerList[]:
+          if
+              it.connected and
+              # avoid negative score peers
+              it.score >= 0.0 and
+              it notin currentMesh and
+              # don't pick explicit peers
+              it.peerId notin g.parameters.directPeers and
+              # and avoid peers we are backing off
+              it.peerId notin backingOff:
+            result.add(it)
+    var candidates =
+      try: getCandidates(g.mesh[topic])
```

Comment:
You shouldn't get `KeyError` if using `withValue`, as a consequence I don't think you need the `getCandidates` closure here.

Reply:
Ah, I see that you're accessing `g.mesh[topic]` which will throw, but a better approach is to use the two body `withValue`:

```nim
  ##   table.withValue(key, value) do:
  ##     # block is executed only if ``key`` in ``t``
  ##     value.name = "username"
  ##     value.uid = 1000
  ##   do:
  ##     # block is executed when ``key`` not in ``t``
  ##     raise newException(KeyError, "Key not found")
```

Reply:
The goal of the closure is to create a "lent variable", but since that's only possible in proc parameters, we have to create a closure
Not sure if withValue will work here (ie, not copy the seqs), I'll try

Reply:
`withValue` will avoid copies.

Reply:
Switched to withValue
Since it works with pointer, I switched everything to pointer, which is less safe, but avoids the closure :shrug: